generate failed: Scanner SubTypesScanner was not configured

When building the following project with:
- maven 3.8.4 
- java version: 11.0.13, vendor: Red Hat, Inc.

```xml
<?xml version="1.0"?>
<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd" 
    xmlns="http://maven.apache.org/POM/4.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">

    <modelVersion>4.0.0</modelVersion>
    <groupId>com.github.lburgazzoli</groupId>
    <artifactId>schema</artifactId>
    <version>0.1</version>

    <packaging>jar</packaging>

    <properties>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
        <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>

        <maven.compiler.target>11</maven.compiler.target>
        <maven.compiler.source>11</maven.compiler.source>
        <maven.compiler.parameters>true</maven.compiler.parameters>
    </properties>

    <dependencies>
        <dependency>
            <groupId>com.slack.api</groupId>
            <artifactId>slack-api-model</artifactId>
            <version>1.8.1</version>
        </dependency>
    </dependencies>

    <build>
        <plugins>
            <plugin>
                <groupId>com.github.victools</groupId>
                <artifactId>jsonschema-maven-plugin</artifactId>
                <version>4.21.0</version>
                <executions>
                    <execution>
                        <goals>
                            <goal>generate</goal>
                        </goals>
                    </execution>
                </executions>
                <configuration>
                    <classNames>
                        <className>com.slack.api.model.Message</className>
                    </classNames>
                </configuration>
            </plugin>
        </plugins>
    </build>

</project>
```

I got the following error:

```
➜ mvn -V package
Apache Maven 3.8.4 (9b656c72d54e5bacbed989b64718c159fe39b537)
Maven home: /home/luca/.local/sfw/tools/sdkman/candidates/maven/current
Java version: 11.0.13, vendor: Red Hat, Inc., runtime: /usr/lib/jvm/java-11-openjdk-11.0.13.0.8-2.fc35.x86_64
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "5.15.12-200.fc35.x86_64", arch: "amd64", family: "unix"
[INFO] Scanning for projects...
[INFO] 
[INFO] -------------------< com.github.lburgazzoli:schema >--------------------
[INFO] Building schema 0.1
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ schema ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 1 resource
[INFO] 
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ schema ---
[INFO] No sources to compile
[INFO] 
[INFO] --- jsonschema-maven-plugin:4.21.0:generate (default) @ schema ---
[INFO] Generating JSON Schema for <className>com.slack.api.model.Message</className>
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.737 s
[INFO] Finished at: 2022-01-07T22:07:28+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.github.victools:jsonschema-maven-plugin:4.21.0:generate (default) on project schema: Execution default of goal com.github.victools:jsonschema-maven-plugin:4.21.0:generate failed: Scanner SubTypesScanner was not configured -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException

```

Extracted from #221

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -12,6 +12,13 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 #### Dependency Update
 - Replace `log4j` test dependency with `logback` (still only test dependency)
 
+### `jsonschema-maven-plugin`
+#### Fixed
+- Enable usage under JDK11 (by adjusting usage of `reflections`)
+
+#### Dependency Update
+- Update `reflections` runtime dependency from 0.9.12 to 0.10.2
+
 ## [4.21.0] - 2021-12-03
 ### `jsonschema-generator`
 #### Fixed

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -14,7 +14,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### `jsonschema-maven-plugin`
 #### Fixed
-- Enable usage under JDK11 (by adjusting usage of `reflections`)
+- Enable usage under JDK11 (by adjusting usage of `reflections`, in order to allow finding classes in dependencies/jar)
 
 #### Dependency Update
 - Update `reflections` runtime dependency from 0.9.12 to 0.10.2

--- a/jsonschema-maven-plugin/README.md
+++ b/jsonschema-maven-plugin/README.md
@@ -101,6 +101,7 @@ The version of JSON Schema that is to be used can be configured with the `<schem
 </configuration>
 ```
 Allowed values are (as per `com.github.victools.jsonschema.generator.SchemaVersion`):
+- `DRAFT_2020_12`
 - `DRAFT_2019_09`
 - `DRAFT_7`
 - `DRAFT_6`

--- a/jsonschema-maven-plugin/src/main/java/com/github/victools/jsonschema/plugin/maven/SchemaGeneratorMojo.java
+++ b/jsonschema-maven-plugin/src/main/java/com/github/victools/jsonschema/plugin/maven/SchemaGeneratorMojo.java
@@ -141,7 +141,7 @@ public class SchemaGeneratorMojo extends AbstractMojo {
     /**
      * The classloader used for loading generator modules and classes.
      */
-    private ClassLoader classLoader = null;
+    private URLClassLoader classLoader = null;
 
     /**
      * The list of all the classes on the classpath.
@@ -230,6 +230,7 @@ public class SchemaGeneratorMojo extends AbstractMojo {
             Scanner subTypeScanner = Scanners.SubTypes.filterResultsBy(c -> true);
             Reflections reflections = new Reflections(new ConfigurationBuilder()
                     .forPackage("", this.getClassLoader())
+                    .addUrls(this.getClassLoader().getURLs())
                     .addScanners(subTypeScanner));
             Stream<PotentialSchemaClass> allTypesStream = reflections.getAll(subTypeScanner)
                     .stream()
@@ -417,7 +418,7 @@ public class SchemaGeneratorMojo extends AbstractMojo {
      *
      * @return The classloader
      */
-    private ClassLoader getClassLoader() {
+    private URLClassLoader getClassLoader() {
         if (this.classLoader == null) {
             List<String> runtimeClasspathElements = null;
             try {

--- a/jsonschema-maven-plugin/src/main/java/com/github/victools/jsonschema/plugin/maven/SchemaGeneratorMojo.java
+++ b/jsonschema-maven-plugin/src/main/java/com/github/victools/jsonschema/plugin/maven/SchemaGeneratorMojo.java
@@ -228,10 +228,14 @@ public class SchemaGeneratorMojo extends AbstractMojo {
     private List<PotentialSchemaClass> getAllClassNames() {
         if (this.allTypes == null) {
             Scanner subTypeScanner = Scanners.SubTypes.filterResultsBy(c -> true);
-            Reflections reflections = new Reflections(new ConfigurationBuilder()
-                    .forPackage("", this.getClassLoader())
-                    .addUrls(this.getClassLoader().getURLs())
-                    .addScanners(subTypeScanner));
+            URLClassLoader urlClassLoader = this.getClassLoader();
+            ConfigurationBuilder configBuilder = new ConfigurationBuilder()
+                    .forPackage("", urlClassLoader)
+                    .addScanners(subTypeScanner);
+            if (urlClassLoader != null) {
+                configBuilder.addUrls(urlClassLoader.getURLs());
+            }
+            Reflections reflections = new Reflections(configBuilder);
             Stream<PotentialSchemaClass> allTypesStream = reflections.getAll(subTypeScanner)
                     .stream()
                     .map(PotentialSchemaClass::new);

--- a/jsonschema-maven-plugin/src/main/java/com/github/victools/jsonschema/plugin/maven/SchemaGeneratorMojo.java
+++ b/jsonschema-maven-plugin/src/main/java/com/github/victools/jsonschema/plugin/maven/SchemaGeneratorMojo.java
@@ -110,7 +110,7 @@ public class SchemaGeneratorMojo extends AbstractMojo {
     private String schemaFileName;
 
     /**
-     * The schema version to be used: DRAFT_6, DRAFT_7 or DRAFT_2019_09.
+     * The schema version to be used: DRAFT_6, DRAFT_7, DRAFT_2019_09 or DRAFT_2020_12.
      */
     @Parameter(property = "schemaVersion", defaultValue = "DRAFT_7")
     private SchemaVersion schemaVersion;

--- a/jsonschema-maven-plugin/src/main/java/com/github/victools/jsonschema/plugin/maven/SchemaGeneratorMojo.java
+++ b/jsonschema-maven-plugin/src/main/java/com/github/victools/jsonschema/plugin/maven/SchemaGeneratorMojo.java
@@ -59,7 +59,9 @@ import org.apache.maven.plugins.annotations.Parameter;
 import org.apache.maven.plugins.annotations.ResolutionScope;
 import org.apache.maven.project.MavenProject;
 import org.reflections.Reflections;
-import org.reflections.scanners.SubTypesScanner;
+import org.reflections.scanners.Scanner;
+import org.reflections.scanners.Scanners;
+import org.reflections.util.ConfigurationBuilder;
 
 /**
  * Maven plugin for the victools/jsonschema-generator.
@@ -225,8 +227,12 @@ public class SchemaGeneratorMojo extends AbstractMojo {
      */
     private List<PotentialSchemaClass> getAllClassNames() {
         if (this.allTypes == null) {
-            Reflections reflections = new Reflections("", new SubTypesScanner(false), this.getClassLoader());
-            Stream<PotentialSchemaClass> allTypesStream = reflections.getAllTypes().stream()
+            Scanner subTypeScanner = Scanners.SubTypes.filterResultsBy(c -> true);
+            Reflections reflections = new Reflections(new ConfigurationBuilder()
+                    .forPackage("", this.getClassLoader())
+                    .addScanners(subTypeScanner));
+            Stream<PotentialSchemaClass> allTypesStream = reflections.getAll(subTypeScanner)
+                    .stream()
                     .map(PotentialSchemaClass::new);
             if (this.excludeClassNames != null && this.excludeClassNames.length > 0) {
                 Set<Pattern> exclusions = Stream.of(this.excludeClassNames)

--- a/pom.xml
+++ b/pom.xml
@@ -118,7 +118,7 @@
         <version.junitparams>1.1.1</version.junitparams>
         <version.logback>1.2.7</version.logback>
         <version.mockito>2.27.0</version.mockito>
-        <version.reflections>0.9.12</version.reflections>
+        <version.reflections>0.10.2</version.reflections>
         <version.slf4j>1.7.32</version.slf4j>
         <version.swagger-1.5>1.5.22</version.swagger-1.5>
         <version.swagger-2>2.1.2</version.swagger-2>

--- a/slate-docs/source/includes/_main-generator.md
+++ b/slate-docs/source/includes/_main-generator.md
@@ -1,4 +1,4 @@
-The [victools:jsonschema-generator](https://github.com/victools/jsonschema-generator/tree/master/jsonschema-generator) aims at allowing the generation of JSON Schema (Draft 6, Draft 7 or Draft 2019-09) to document Java code.
+The [victools:jsonschema-generator](https://github.com/victools/jsonschema-generator/tree/master/jsonschema-generator) aims at allowing the generation of JSON Schema (Draft 6, Draft 7, Draft 2019-09 or Draft 2020-12) to document Java code.
 This is expressly not limited to _JSON_ but also allows for a Java API to be documented (i.e. including methods and the associated return values).
 
 # Generator – Options

--- a/slate-docs/source/includes/_maven-plugin.md
+++ b/slate-docs/source/includes/_maven-plugin.md
@@ -31,7 +31,7 @@ There are some additional parameters available in the plugin `<configuration>`:
 | --- | --- | --- | --- |
 |  1 | `<schemaFilePath>` | `src/main/resources` | Directory to generate all schemas in |
 |  2 | `<schemaFileName>` | `{0}-schema.json` | Relative path from the `<schemaFilePath>` including the file name pattern. Two placeholders are supported: `{0}` will be replaced with the respective simple class name (e.g. `TypeA`) `{1}` will be replaced with the respective package path (e.g. `com/myOrg/myApp`) in case you want to preserve the original package structure |
-|  3 | `<schemaVersion>` | `DRAFT_7` | JSON Schema version to apply (`DRAFT_6`, `DRAFT_7` or `DRAFT_2019_09`) |
+|  3 | `<schemaVersion>` | `DRAFT_7` | JSON Schema version to apply (`DRAFT_6`, `DRAFT_7`, `DRAFT_2019_09` or `DRAFT_2020_12`) |
 
 ```xml
 <options>

